File: 00004/onp.rb
n = Integer(gets.chomp)

n.times do
    exp = gets.chomp
    optr = []
    opnd = []
    ops = ["(", "+", "-", "*", "/", "^", ")"]
    exp.each_byte do |ch|
        ch = ch.chr

        if ch >= 'a' and ch <= 'z'
            opnd.push(ch)
        end

        if ch == '('
            optr.push(ch)
        end

        if ch == ')'
            while optr.last != '(' do
                opnd.push([opnd.pop, opnd.pop].reverse.join + optr.pop)
            end
            optr.pop
        end

        if ops.slice(1,5).include?(ch)
            while ops.index(ch)<=ops.index(optr.last) do
                opnd.push([opnd.pop, opnd.pop].reverse.join + optr.pop)
            end
            optr.push(ch)
        end
    end
    puts opnd
end

File: spoj/00004/onp.py
import sys

n = int(sys.stdin.readline().strip())
alpha = map(lambda x: chr(x+ord('a')), range(26))
ops = "(+-*/^)"
for c in xrange(n):
    exp = sys.stdin.readline().strip()
    optr = []; opnd = []
    for ch in exp:
        if ch in alpha: opnd.append(ch)
        if ch == "(": optr.append(ch)
        if ch == ")":
            while optr[-1] != "(":
                opnd.append("".join([opnd.pop(), opnd.pop()][::-1]) + optr.pop())
            optr.pop()
        if ch in ops[1:6]:
            while ops.find(ch)<= ops.find(optr[-1]):
                opnd.append("".join([opnd.pop(), opnd.pop()][::-1]) + optr.pop())
            optr.append(ch)
    print opnd[0]
